vendor/bat-native-ledger/src/bat/ledger/internal/uphold/uphold_authorization.cc

```diff
@@ -1,72 +1,67 @@
-/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
  * This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 #include "bat/ledger/internal/uphold/uphold_authorization.h"
 
-#include <utility>
-
-#include "base/json/json_reader.h"
-#include "base/strings/stringprintf.h"
-#include "bat/ledger/global_constants.h"
 #include "bat/ledger/internal/common/random_util.h"
 #include "bat/ledger/internal/ledger_impl.h"
 #include "bat/ledger/internal/logging/event_log_keys.h"
 #include "bat/ledger/internal/uphold/uphold_util.h"
 
 using std::placeholders::_1;
 using std::placeholders::_2;
-using std::placeholders::_3;
 
 namespace ledger {
 namespace uphold {
 
-UpholdAuthorization::UpholdAuthorization(LedgerImpl* ledger) :
-    ledger_(ledger),
-    uphold_server_(std::make_unique<endpoint::UpholdServer>(ledger)) {
-}
+UpholdAuthorization::UpholdAuthorization(LedgerImpl* ledger)
+    : ledger_(ledger),
+      uphold_server_(std::make_unique<endpoint::UpholdServer>(ledger)) {}
 
 UpholdAuthorization::~UpholdAuthorization() = default;
 
 void UpholdAuthorization::Authorize(
     const base::flat_map<std::string, std::string>& args,
-    ledger::ExternalWalletAuthorizationCallback callback) {
-  auto wallet = GetWallet(ledger_);
+    ledger::ExternalWalletAuthorizationCallback callback) const {
+  auto uphold_wallet = GetWallet(ledger_);
+  if (!uphold_wallet) {
+    BLOG(0, "The Uphold wallet is null!");
+    return callback(type::Result::LEDGER_ERROR, {});
+  }
 
-  if (!wallet) {
-    BLOG(0, "Wallet is null");
-    callback(type::Result::LEDGER_ERROR, {});
-    return;
+  if (uphold_wallet->status != type::WalletStatus::NOT_CONNECTED &&
```

Comment:
Requesting a re-authorization while being in the `PENDING` or `VERIFIED` state wouldn't really make much sense IMO, since the user already has a valid `access_token` in both cases. I don't even think the user could achieve this on the UI by the way, except for one scenario: if you have two `brave://rewards` tabs open before you would authorize with Uphold, you do the authorization on one of them, then you're actually left with a UI on the other one that does offer the possibility to re-authorize, which is exactly what this check is intended to protect us from. I sent a video on that the other day. 🙂

Reply:
Totally makes sense. If we think that the system should not get into this state in normal operation, then:

- Should we callback with `LEDGER_ERROR` to indicate that the authorization was not successful? (After all, they could be attempting to authorize with a different account or something.)
- Should we log an error when we detect this condition?

Reply:
Logging an error makes sense, sure!
Returning `LEDGER_ERROR` takes us to the error modal:
![error_modal](https://user-images.githubusercontent.com/5390451/123641494-9aa62700-d822-11eb-8128-2b3b42e3d447.png)
But we can do even better, see my [comment](https://github.com/brave/brave-core/pull/9212#issuecomment-869171510).

Reply:
Added
```
BLOG(0, "Attempting to re-authorize in " << uphold_wallet->status << " status!");
return callback(type::Result::LEDGER_ERROR, {});
```
for now.
Wrapping this is up in https://github.com/brave/brave-browser/issues/16652.